### ISSUE TITLE
Some strings not translatable in M2.4.1

### ISSUE THREAD

**Issue by siimm** [magento/magento2#31849]
### Preconditions (*)
1. Clean Magento 2.4.1 installation
1. 2.4-develop

### Steps to reproduce (*)
1. Add a product to cart
1. Navigate to cart page
1. Open admin interface, find the product, save it as "Out of Stock"
1. Refresh cart page, see error message: **There are no source items with the in stock status**
1. Find the translation for the message in <magento install dir>vendor/magento/module-catalog-inventory/i18n/en_US.csv
1. Change the result string to be something else
1. Clean "translate" cache
1. Refresh the cart page and witness the previously mentioned string being not translated to value we changed it at point 6

### Expected result (*)
1. Error messages are translated

### Actual result (*)
1. Error messages are not translated - can be verified even with default en_US language - change the translation for the error message

### NOTES
The **steps to reproduce** part only highlights one specific string, however this seems to be a more general problem:

Due to update of design (and translations) loading in https://github.com/magento/magento2/commit/b2a035d93c98487b69bbcd6b719e79ddf8e39574
(loading design and translations was moved from beforeDispatch to beforeExecute)

However there are several plugins in M2 ecosystem, which come before actual execute call, \Magento\InventorySales\Model\IsProductSalableForRequestedQtyCondition\IsProductSalableForRequestedQtyConditionChain is one of the many examples (on quote loading, validation is triggered).

Verified this in clean Magento 2.4.1 - the error phrase is never translated, as renderers for Phrase is set only in \Magento\Theme\Plugin\LoadDesignPlugin::beforeExecute

This happens due to renderer not being set to \Magento\Framework\Phrase, and if it is used earlier than the renderers are set - the original string gets returned, as it uses the RendererPlaceholder class to render them.

With current state, a lot of strings cannot be translated, the only imaginable workaround is to move design and translations loading back to beforeDispatch method. (Moving just the translation loading to beforeDispatch won't help, if the translations are in theme.)

## 
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- < ] Severity: **S0** *- Affects critical data or functionality and leaves users without workaround.*
- [ ] Severity: **S1** *- Affects critical data or functionality and forces users to employ a workaround.*
- [x] Severity: **S2** *- Affects non-critical data or functionality and forces users to employ a workaround.*
- [ ] Severity: **S3** *- Affects non-critical data or functionality and does not force users to employ a workaround.*
- [ > Severity: **S4** *- Affects aesthetics, professional look and feel, “quality” or “usability”.*

**Comment by ihor-sviziev** [comment]
@sidolov @gabrieldagama, I believe this is a significant issue. Could you take a look at it and set priority on it?

**Comment by engcom-Alfa** [comment]
@siimm .
We are able to reproduce this issue on clean Magento 2.4-develop + **Inventory**

**Manual testing scenario:** 

1. Add a product to cart
2. Navigate to cart page
3. Open admin interface, find the product, save it as "Out of Stock"
4. Refresh cart page, see error message: **There are no source items with the in stock status**
5. Find the translation for the message in <magento install dir>vendor/magento/module-catalog-inventory/i18n/en_US.csv
6. Change the result string to be something else
7. Clean "translate" cache
8. Refresh the cart page and witness the previously mentioned string being not translated to value we changed it at point 6

**Actual Result:** :heavy_multiplication_x: Error messages are not translated

![Screenshot from 2021-01-27 15-32-17](https://user-images.githubusercontent.com/51679138/105998530-684a9f00-60b5-11eb-85ef-9f22c3783c8c.png)

but we are not able to reproduce it on clean Magento 2.4-develop:

**Manual testing scenario:** 

1. Add a product to cart
2. Navigate to cart page
3. Open admin interface, find the product, save it as "Out of Stock"
4. Refresh cart page, see error message: **This product is out of stock**
5. Find the translation for the message in <magento install dir>vendor/magento/module-catalog-inventory/i18n/en_US.csv
6. Change the result string to be something else
7. Clean "translate" cache
8. Refresh the cart page and witness the previously mentioned string being not translated to value we changed it at point 6

**Actual Result:** :heavy_check_mark: Error messages are translated

![Screenshot from 2021-01-27 15-22-37](https://user-images.githubusercontent.com/51679138/105998835-b3fd4880-60b5-11eb-9cb4-639f8934a278.png)

@siimm @ihor-sviziev Can we assume that the root of the problem is in the **Inventory**? Is this problem not related to MSI?
Thanks!

**Comment by siimm** [comment]
hi @engcom-Alfa 

The inventory message served as an example, what I wanted to highlight was a more general problem with the architectural design of the current solution.
I am convinced that if I go looking for next examples, I will find them.

A brief look into predispatch plugins already reveals bunch of points to check, not speaking of classes which are used in those plugins that also have translations.
```
Siims-MBP :: m2clean/vendor/magento » grep -l -r "n beforeDispatch" .|xargs grep "__("
./module-invitation/Controller/Customer/Account/Plugin.php:            throw new NotFoundException(__('Page not found.'));
./module-wishlist/Controller/Index/Plugin.php:                $this->messageManager->addErrorMessage(__('You must login or register to add items to your wishlist.'));
./module-wishlist/Controller/Index/Plugin.php:            throw new NotFoundException(__('Page not found.'));
./module-multiple-wishlist/Controller/Index/Plugin.php:            throw new NotFoundException(__('Page not found.'));
./module-store/App/Action/Plugin/Context.php:                ? __($previousException->getMessage())
./module-store/App/Action/Plugin/Context.php:                : __('Invalid store requested.'),
./module-deploy/Model/Plugin/ConfigChangeDetector.php:                __(
```

If you like, I can come up with next examples.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Alfa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40654` were created

**Issue Available**: @engcom-Alfa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Alfa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40654` were created

**Issue Available**: @engcom-Alfa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ansari-krish** [comment]
@magento I am working on this

**Comment by mrtuvn** [comment]
CC @lbajsarowicz @Vinai @vkublytskyi

**Comment by mrtuvn** [comment]
@engcom-Alfa can you test again after this PR commit merged
Seem problem fixed after this PR https://github.com/magento/magento2/pull/31970

**Comment by engcom-Alfa** [comment]
Hi @mrtuvn. I rechecked it and the issue is still reproducible on a fresh 2.4-develop.

*Manual testing scenario:** 

1. Add a product to cart
2. Navigate to cart page
3. Open admin interface, find the product, save it as "Out of Stock"
4. Refresh cart page, see error message: **There are no source items with the in stock status**
5. Find the translation for the message in <magento install dir>vendor/magento/module-catalog-inventory/i18n/en_US.csv
6. Change the result string to be something else
7. Clean "translate" cache
8. Refresh the cart page and witness the previously mentioned string being not translated to value we changed it at point 6

**Actual Result:** :heavy_multiplication_x: Error messages are not translated

![Screenshot from 2021-01-27 15-32-17](https://user-images.githubusercontent.com/51679138/105998530-684a9f00-60b5-11eb-85ef-9f22c3783c8c.png)

**Comment by mrtuvn** [comment]
"There are no source items with the in stock status" Seem from another place
As i remember it's from module https://github.com/magento/inventory/blob/1.2-develop/InventorySales/i18n/en_US.csv
Seem path change incorrect right ?

**Comment by engcom-Alfa** [comment]
@mrtuvn For now, for example, if we take the phrase _**'You must login or register to add items to your wishlist.'**_  from the [comment](https://github.com/magento/magento2/issues/31849#issuecomment-768298340) error message is translated:

![Screenshot from 2021-02-19 13-02-53](https://user-images.githubusercontent.com/51679138/108496508-d1879180-72b2-11eb-91a2-174bbd8a552f.png)

but if we check the scenario from the issue with Inventory the error message is not translated. Other phrases from the Inventory are translated correctly.
Can we assume that this problem belongs to the Inventory?

**Comment by mrtuvn** [comment]
It's possible but I'm not sure when inventory is separated. Maybe we need someone active on inventory repo confirm that

**Comment by mrtuvn** [comment]
Seem some old controllers in codebase (inventory or EE) still use old approach with beforeDispatch (By interact extend deprecated Action Controller)
So some messages around here may not able translated correctly

**Comment by gabrieldagama** [comment]
Hi @mrtuvn @engcom-Alfa I believe the problem mentioned previously on this issue was fixed. The problem with the `There are no source items with the in stock status` string not being translatable is likely a different one. It seems to be related to the way these error messages are handled.

I believe this issue wasn't reproducible anymore and we should close it. We can open a new issue to handle the error messages that come from \Magento\Quote\Model\Cart\AddProductsToCart as on the example from @engcom-Alfa.

We should also close your PR @mrtuvn. Please let me know if it makes sense.

Thanks.

**Comment by mrtuvn** [comment]
I'm not sure there are plenty places can show message when error happen! We may not check all. Let's author recheck and decide.

**Comment by gabrieldagama** [comment]
@siimm can you confirm if this error still reproducible with the string `Some of the products are out of stock.` as described when the issue was opened?

**Comment by siimm** [comment]
@gabrieldagama I will try and get back to you

**Comment by mrtuvn** [comment]
Note this patch not available in 2.4.1 or 2.4.2. You need to reproduce issue with github 2.4-develop
